src/app/profiles/profile-detail/profile-detail.component.ts

```diff
@@ -123,37 +107,46 @@ export class ProfileDetailComponent implements OnInit {
     })
   }
 
-  generateRandomPasswordChange (value: boolean): void {
-    if (value) {
-      this.profileForm.controls.amtPassword.disable()
-      this.profileForm.controls.amtPassword.setValue(null)
-      this.profileForm.controls.amtPassword.clearValidators()
-      this.profileForm.controls.amtPassword.setValue(null)
-      this.profileForm.controls.passwordLength.setValue(8)
-      this.profileForm.controls.passwordLength.enable()
-    } else {
-      this.profileForm.controls.amtPassword.enable()
-      this.profileForm.controls.amtPassword.setValidators(Validators.required)
-      this.profileForm.controls.passwordLength.disable()
-      this.profileForm.controls.passwordLength.setValue(null)
-      this.profileForm.controls.passwordLength.clearValidators()
+  generateRandomPassword (controlName: string, length: number): void {
+    const charset: RegExp = /[a-zA-Z0-9!$%]/
+    const requirements: RegExp[] = [/[a-z]/, /[A-Z]/, /[0-9]/, /[!$%]/]
+    const bit = new Uint8Array(1)
+    let char: string = ''
+    let password: string = ''
+    let searching: boolean = true
+
+    while (searching) {
+      for (let i = 0; i < length; i++) {
+        char = ''
+        while (!charset.test(char)) {
+          window.crypto.getRandomValues(bit)
+          char = String.fromCharCode(bit[0])
+        }
+        password += char
+      }
+
+      searching = false
+
+      for (let i = 0; i < requirements.length; i++) {
+        if (!requirements[i].test(password)) {
+          searching = true
+          password = ''
+        }
+      }
     }
+
+    this.profileForm.controls[controlName].setValue(password)
   }
 
-  generateRandomMEBxPasswordChange (value: boolean): void {
-    if (value) {
-      this.profileForm.controls.mebxPassword.disable()
-      this.profileForm.controls.mebxPassword.setValue(null)
-      this.profileForm.controls.mebxPassword.clearValidators()
-      this.profileForm.controls.mebxPassword.setValue(null)
-      this.profileForm.controls.mebxPasswordLength.setValue(8)
-      this.profileForm.controls.mebxPasswordLength.enable()
-    } else if (this.profileForm.controls.activation.value === Constants.ACMActivate) {
-      this.profileForm.controls.mebxPassword.enable()
-      this.profileForm.controls.mebxPassword.setValidators(Validators.required)
-      this.profileForm.controls.mebxPasswordLength.disable()
-      this.profileForm.controls.mebxPasswordLength.setValue(null)
-      this.profileForm.controls.mebxPasswordLength.clearValidators()
+  togglePassVisibility (passId: string, iconId: string): void {
+    if (document.getElementById(passId)?.getAttribute('type') === 'password') {
```

Comment:
This should be much simplere. we shouldn't be doing DOM manipulation here. I think the better way to do this is to bind the type property of the password field and toggle it between password/text

Reply:
I agree, I tried to make it clean originally but I kept getting errors